### PR TITLE
WIP: SCA upgrade

### PR DESCRIPTION
**Pull request by NorfairKing**

### CODE REVIEW

**Comment by dmjio** [comment]
@NorfairKing looks good, seems the `FromJSON` needs to be fleshed out.

**Comment by NorfairKing** [comment]
@dmjio Yes, everything about this is not done. Is this really the only set of endpoints that needs to be finished?

**Comment by dmjio** [comment]
@NorfairKing seems like we need to facilitate this payment workflow

image

**Comment by dmjio** [comment]
We might also need `SetupIntents`

https://stripe.com/docs/api/setup_intents

**Comment by NorfairKing** [comment]
We'll also need the new events then, because that's how you do the last part.

**Comment by emhoracek** [comment]
My project also needs this functionality. Just let me know what I can help with!

**Comment by NorfairKing** [comment]
@emhoracek It would be great if you could help finish the pull request because I don't have much time.

**Comment by emhoracek** [comment]
@dmjio It looks like `SetupIntents` is only needed if you want to save cards for later without charging them, so that could be a separate PR.

Screen Shot 2019-09-10 at 12 06 39 PM

**Comment by NorfairKing** [comment]
To anyone reading this: I have no unpushed local work, please help out because I don't really have the time to complete this.

**Comment by NorfairKing** [comment]
@dmjio Could you have a look and tell me what else needs to be done for payment intents?

**Comment by sigrlami** [comment]
@NorfairKing what's the status of this. Did you manage to build `PaymentIntent` and related workflow successfully? Thanks

**Comment by NorfairKing** [comment]
@sigrlami Yes, but I don't have the funds/time to finish this PR. I'm just using my own branch for the moment.

**Comment by 2piix** [comment]
What's missing for this PR?  I need Sessions support, and I have some free time to try this week.

**Comment by dmjio** [comment]
@NorfairKing

**Comment by NorfairKing** [comment]
@2piix Go for it. I won't have the resources to finish this PR. I thought I did something on the topic of sessions though. Have a look.

File: stripe-core/src/Web/Stripe/Types.hs [hunk 1]
```diff
@@ -535,13 +512,98 @@ data DefaultCard = DefaultCard { getDefaultCard :: CardId }
...
+
+data Session = Session {
+      sessionId                         :: SessionId
+    , sessionCancelUrl :: CancelUrl
+    , sessionSuccessUrl :: SuccessUrl
+    , sessionLivemode :: Bool
+    , sessionClientReferenceId :: Maybe ClientReferenceId
+    , sessionCustomerEmail :: Maybe CustomerEmail
+    , sessionBillingAddressCollection :: Maybe TODO
+    , sessionDisplayItems :: Maybe [TODO]
```
Comment (jezen) [inline stripe-core/src/Web/Stripe/Types.hs:526]: @NorfairKing Do you remember where you got `sessionDisplayItems` from? I can't see anything about display items in [the API reference][0].

[0]: https://stripe.com/docs/api/checkout/sessions

**Comment by emhoracek** [comment]
Position Dev managed to take this branch and add a few more types and endpoints and get our client set up for SCA - thanks @NorfairKing for getting this started.

I'm going to try to finish the upgrade (currently still has a lot of `TODO` types for the stuff we didn't end up needing immediately) and contribute it back - I'll open a PR so you can see what we've added so far. It's a bit messy since we were trying to get it done by EOY.

**Comment by NorfairKing** [comment]
Closing this.

* It's subsumed by #135
* I can't finish it
* I don't use it anymore.

**Comment by dmjio** [comment]
@NorfairKing should we deprecate in favor of https://github.com/Haskell-OpenAPI-Code-Generator/Stripe-Haskell-Library ?

**Comment by NorfairKing** [comment]
> @NorfairKing should we deprecate in favor of https://github.com/Haskell-OpenAPI-Code-Generator/Stripe-Haskell-Library ?

I generate a custom `stripe-client` for every use-case, because that library takes ages to compile, with this: https://github.com/Haskell-OpenAPI-Code-Generator/Haskell-OpenAPI-Client-Code-Generator

I wouldn't necessarily say deprecate it, but this library is out-of-date for my use-case and I find it more sustainable to generate code instead.

**Pull request closed without merging**